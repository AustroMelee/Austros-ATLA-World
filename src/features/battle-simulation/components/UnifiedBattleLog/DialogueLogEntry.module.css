.container {
  padding: 2px 0;
  display: flex;
}
.left {
  justify-content: flex-start;
  text-align: left;
}
.right {
  justify-content: flex-end;
  text-align: right;
}
.center {
  justify-content: center;
  text-align: center;
}
.actor {
  font-weight: bold;
  margin-right: 4px;
}
.text {
  font-family: inherit;
} 