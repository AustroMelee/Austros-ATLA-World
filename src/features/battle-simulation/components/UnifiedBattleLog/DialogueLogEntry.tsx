interface DialogueLogEntryProps {
  actor: string;
  text: string;
  align: 'left' | 'right' | 'center';
  color?: string;
  turn?: number;
}

export default function DialogueLogEntry({ actor, text, align, color, turn }: DialogueLogEntryProps) {
  return (
    <div
      style={{
        textAlign: align,
        margin: '0.5em 0',
        padding: '1em',
        borderRadius: 8,
        background: color || (align === 'left' ? '#1e293b' : '#7f1d1d'),
        color: '#fff',
        maxWidth: 500,
        marginLeft: align === 'left' ? 0 : 'auto',
        marginRight: align === 'right' ? 0 : 'auto',
        boxShadow: '0 2px 8px rgba(0,0,0,0.08)'
      }}
    >
      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start' }}>
        <div style={{ fontWeight: 900, fontSize: 20, marginBottom: 8 }}>{actor}</div>
        {typeof turn === 'number' && (
          <div style={{ fontSize: 14, color: '#cbd5e1', fontWeight: 600, marginLeft: 12 }}>T{turn}</div>
        )}
      </div>
      <div style={{ fontSize: 16 }}>{text}</div>
    </div>
  );
} 