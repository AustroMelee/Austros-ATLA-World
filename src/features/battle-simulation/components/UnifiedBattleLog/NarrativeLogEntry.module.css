.container {
  padding: 2px 0;
  display: flex;
}
.left {
  justify-content: flex-start;
  text-align: left;
}
.center {
  justify-content: center;
  text-align: center;
}
.text {
  font-family: inherit;
} 