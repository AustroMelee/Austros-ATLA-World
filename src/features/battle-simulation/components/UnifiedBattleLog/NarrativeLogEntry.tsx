interface NarrativeLogEntryProps {
  text: string;
  align?: 'left' | 'center' | 'right';
  actor?: string;
  turn?: number;
  color?: string;
}

export default function NarrativeLogEntry({ text, align = "center", actor, turn, color }: NarrativeLogEntryProps) {
  return (
    <div
      style={{
        textAlign: align,
        margin: '0.5em 0',
        padding: '1em',
        borderRadius: 8,
        background: color || '#334155',
        color: '#fff',
        maxWidth: 500,
        marginLeft: align === 'left' ? 0 : 'auto',
        marginRight: align === 'right' ? 0 : 'auto',
        boxShadow: '0 2px 8px rgba(0,0,0,0.08)'
      }}
    >
      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start' }}>
        {actor && <div style={{ fontWeight: 900, fontSize: 20, marginBottom: 8 }}>{actor}</div>}
        {typeof turn === 'number' && (
          <div style={{ fontSize: 14, color: '#cbd5e1', fontWeight: 600, marginLeft: 12 }}>T{turn}</div>
        )}
      </div>
      <div style={{ fontSize: 16, fontStyle: 'italic', whiteSpace: 'pre-line', wordBreak: 'break-word' }}>{text}</div>
    </div>
  );
} 