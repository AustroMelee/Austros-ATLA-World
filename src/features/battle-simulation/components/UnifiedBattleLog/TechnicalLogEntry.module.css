.container {
  padding: 2px 0;
}
.text {
  font-family: monospace;
  font-size: 0.95em;
} 