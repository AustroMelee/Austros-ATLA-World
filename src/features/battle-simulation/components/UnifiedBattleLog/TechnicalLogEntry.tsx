import React from 'react';
import styles from './TechnicalLogEntry.module.css';

interface TechnicalLogEntryProps {
  text: string;
  align?: 'center';
  title?: string;
  turn?: number;
}

export default function TechnicalLogEntry({ text, align = 'center', title = 'System', turn }: TechnicalLogEntryProps) {
  return (
    <div
      style={{
        textAlign: align,
        margin: '0.5em 0',
        padding: '1em',
        borderRadius: 8,
        background: '#22223b',
        color: '#fff',
        maxWidth: 500,
        marginLeft: 'auto',
        marginRight: 'auto',
        boxShadow: '0 2px 8px rgba(0,0,0,0.08)'
      }}
    >
      <div style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'flex-start' }}>
        <div style={{ fontWeight: 900, fontSize: 20, marginBottom: 8 }}>{title}</div>
        {typeof turn === 'number' && (
          <div style={{ fontSize: 14, color: '#cbd5e1', fontWeight: 600, marginLeft: 12 }}>T{turn}</div>
        )}
      </div>
      <div style={{ fontSize: 16 }}>{text}</div>
    </div>
  );
} 