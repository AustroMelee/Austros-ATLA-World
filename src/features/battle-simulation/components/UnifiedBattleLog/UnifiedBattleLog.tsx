// CONTEXT: Unified Battle Log
// RESPONSIBILITY: Single log component with tabs for narrative and AI logs
// 
// ‚ö†Ô∏è CRITICAL REQUIREMENT: Turn 1 logs MUST ALWAYS be visible by default
// No matter how many updates or changes we make, users MUST be able to see logs from T1
// This is essential for battle analysis and debugging - never hide early turns!
//
import React, { useState } from 'react';
import { BattleLogEntry, AILogEntry, BattleCharacter } from '../../types';
import styles from './UnifiedBattleLog.module.css';

interface UnifiedBattleLogProps {
  battleLog: BattleLogEntry[];
  aiLog: AILogEntry[];
  maxEntries?: number;
  participants?: BattleCharacter[];
}

type LogTab = 'narrative' | 'ai';

/**
 * @description Unified battle log with tabs for narrative and AI logs.
 */
export const UnifiedBattleLog: React.FC<UnifiedBattleLogProps> = ({
  battleLog,
  aiLog,
  maxEntries = 15,
  participants = []
}) => {
  // ‚ö†Ô∏è CRITICAL: Always start with showAllEntries = true to ensure T1 logs are visible
  // This is a hard requirement - users must always see the complete battle log by default
  const [activeTab, setActiveTab] = useState<LogTab>('narrative');
  const [copied, setCopied] = useState(false);
  const [showAllEntries, setShowAllEntries] = useState(true); // ALWAYS TRUE BY DEFAULT
  const [turnFilter, setTurnFilter] = useState<number | null>(null);

  /**
   * @description Gets the appropriate icon for a battle event.
   */
  const getEventIcon = (entry: BattleLogEntry): string => {
    // Opening sequence entries get a special icon
    if (entry.turn >= 1 && entry.turn <= 6) {
      return 'üé≠';
    }
    
    if (entry.type === 'VICTORY') return 'üèÜ';
    if (entry.type === 'DESPERATION') return '‚ö°';
    if (entry.type === 'FINISHER') return 'üí•';
    if (entry.type === 'NARRATIVE') return 'üí≠';
    if (entry.meta?.crit === true) return 'üí•';
    if (entry.meta?.finisher === true) return 'üî•';
    if (entry.meta?.rest === true) return 'üîÑ';
    if (entry.meta?.heal === true) return 'üíö';
    if (entry.meta?.piercing === true) return '‚öîÔ∏è';
    if (entry.damage && entry.damage > 15) return 'üí¢';
    if (entry.damage && entry.damage > 0) return '‚öîÔ∏è';
    if (entry.abilityType === 'defense_buff') return 'üõ°Ô∏è';
    return 'üé≠';
  };

  /**
   * @description Gets CSS classes for a battle log entry.
   */
  const getBattleEntryClasses = (entry: BattleLogEntry): string => {
    const classes = [styles.logEntry];
    
    // Opening sequence classes (turns 1-6)
    if (entry.turn >= 1 && entry.turn <= 6) {
      classes.push(styles.openingSequence);
    }
    
    // Type-based classes
    classes.push(styles[entry.type.toLowerCase()]);
    
    // Meta-based classes
    if (entry.meta?.crit === true) classes.push(styles.critical);
    if (entry.meta?.finisher === true) classes.push(styles.finisher);
    if (entry.meta?.rest === true) classes.push(styles.rest);
    if (entry.meta?.heal === true) classes.push(styles.heal);
    if (entry.meta?.piercing === true) classes.push(styles.piercing);
    if (entry.meta?.desperation === true) classes.push(styles.desperation);
    
    // Damage-based classes
    if (entry.damage) {
      if (entry.damage > 20) classes.push(styles.massiveDamage);
      else if (entry.damage > 10) classes.push(styles.highDamage);
      else if (entry.damage > 0) classes.push(styles.normalDamage);
    }
    
    // Highlight state change
    if (entry.type === 'STATUS' && entry.action === 'State Change') {
      classes.push(styles.stateChange);
      if (entry.meta?.newControlState === 'Compromised') classes.push(styles.compromised);
      if (entry.meta?.newControlState === 'Defeated') classes.push(styles.defeated);
    }
    
    return classes.join(' ');
  };

  /**
   * @description Formats the battle log entry text.
   */
  const formatBattleEntryText = (entry: BattleLogEntry): React.ReactNode => {
    const icon = getEventIcon(entry);
    // Show disruption mechanics if present
    if (entry.meta?.controlShift !== undefined || entry.meta?.stabilityChange !== undefined) {
      return (
        <>
          {icon} <b>{entry.action}</b>: {entry.narrative}
          {entry.meta.controlShift !== undefined ? (
            <span className={styles.controlShift}> [Control Shift: {entry.meta.controlShift}]</span>
          ) : null}
          {entry.meta.stabilityChange !== undefined ? (
            <span className={styles.stabilityChange}> [Stability: -{entry.meta.stabilityChange}]</span>
          ) : null}
          {entry.meta.newControlState ? (
            <span className={styles.newControlState}> [State: {entry.meta.newControlState}]</span>
          ) : null}
        </>
      );
    }
    // Highlight state change log entries
    if (entry.type === 'STATUS' && entry.action === 'State Change') {
      return (
        <>
          {icon} <b>{entry.actor}</b>: <span className={styles.stateChangeBadge}>{entry.narrative}</span>
          {entry.meta?.newControlState ? (
            <span className={styles.newControlState}> [State: {entry.meta.newControlState}]</span>
          ) : null}
        </>
      );
    }
    // Fallback to old logic
    const baseText = entry.narrative || entry.result || entry.action;
    
    // Add damage highlight
    if (entry.damage) {
      const damageText = ` (${entry.damage} damage)`;
      const parts = baseText.split(damageText);
      
      if (parts.length > 1) {
        return (
          <>
            {icon} {parts[0]}
            <span className={styles.damageHighlight}>{damageText}</span>
            {parts[1]}
          </>
        );
      }
    }
    
    // Add critical hit highlight
    if (entry.meta?.crit === true) {
      const critText = ' (CRITICAL!)';
      const parts = baseText.split(critText);
      
      if (parts.length > 1) {
        return (
          <>
            {icon} {parts[0]}
            <span className={styles.criticalHighlight}>{critText}</span>
            {parts[1]}
          </>
        );
      }
    }
    
    return (
      <>
        {icon} {baseText}
      </>
    );
  };

  /**
   * @description Gets the timestamp display for an entry.
   */
  const getTimestamp = (timestamp: number): string => {
    const date = new Date(timestamp);
    return date.toLocaleTimeString('en-US', { 
      hour12: false, 
      hour: '2-digit', 
      minute: '2-digit',
      second: '2-digit'
    });
  };

  /**
   * @description Renders the narrative/battle log tab.
   * ‚ö†Ô∏è CRITICAL: This function MUST always show T1 logs by default
   * The showAllEntries state should default to true to ensure complete visibility
   */
  const renderNarrativeTab = () => {
    // ‚ö†Ô∏è CRITICAL: Always show all entries by default to ensure T1 logs are visible
    // Only slice if explicitly requested by user (showAllEntries = false)
    let limitedEntries = showAllEntries ? battleLog : battleLog.slice(-maxEntries);
    
    // Apply turn filter if set
    if (turnFilter !== null) {
      limitedEntries = limitedEntries.filter(entry => entry.turn === turnFilter);
    }

    // DEV ONLY: Duplicate key detector
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const viteEnv = (import.meta as any).env;
    if ((viteEnv && viteEnv.MODE === 'development') || (!viteEnv || !viteEnv.MODE)) {
      const seen = new Set<string>();
      for (const entry of limitedEntries) {
        if (seen.has(entry.id)) {
          // eslint-disable-next-line no-console
          console.error('DUPLICATE LOG KEY DETECTED:', entry.id, entry);
          // eslint-disable-next-line no-console
          console.log('FULL BATTLE LOG:', limitedEntries);
        }
        seen.add(entry.id);
      }
    }

    return (
      <div className={styles.tabContent}>
        {limitedEntries.length === 0 ? (
          <div className={styles.emptyState}>
            <span className={styles.emptyIcon}>‚öîÔ∏è</span>
            <p>Battle log is empty</p>
          </div>
        ) : (
          limitedEntries.map((entry) => (
            <div key={entry.id} className={getBattleEntryClasses(entry)}>
              <div className={styles.entryHeader}>
                <span className={styles.turnNumber}>T{entry.turn}</span>
                <span className={styles.actor}>{entry.actor}</span>
                <span className={styles.timestamp}>{getTimestamp(entry.timestamp)}</span>
              </div>
              
              <div className={styles.entryContent}>
                {formatBattleEntryText(entry)}
              </div>
              
              {entry.meta && Object.keys(entry.meta).length > 0 && (
                <div className={styles.entryMeta}>
                  {entry.meta.resourceCost !== undefined ? (
                    <span className={styles.resourceCost}>üí† {Number(entry.meta.resourceCost)} chi</span>
                  ) : null}
                  {entry.meta.finisher === true ? (
                    <span className={styles.finisherBadge}>üî• FINISHER</span>
                  ) : null}
                  {entry.meta.desperation === true ? (
                    <span className={styles.desperationBadge}>‚ö° DESPERATION</span>
                  ) : null}
                </div>
              )}
            </div>
          ))
        )}
      </div>
    );
  };

  /**
   * @description Renders the AI log tab.
   */
  const renderAITab = () => {
    let limitedEntries = showAllEntries ? aiLog : aiLog.slice(-maxEntries);
    
    // Apply turn filter if set
    if (turnFilter !== null) {
      limitedEntries = limitedEntries.filter(entry => entry.turn === turnFilter);
    }

    return (
      <div className={styles.tabContent}>
        {limitedEntries.length === 0 ? (
          <div className={styles.emptyState}>
            <span className={styles.emptyIcon}>ü§ñ</span>
            <p>AI log is empty</p>
          </div>
        ) : (
          limitedEntries.map((entry, index) => (
            <div key={`${entry.turn}-${entry.agent}-${index}`} className={`${styles.logEntry} ${styles.aiEntry}`}>
              <div className={styles.entryHeader}>
                <span className={styles.turnNumber}>T{entry.turn}</span>
                <span className={styles.actor}>{entry.agent}</span>
                <span className={styles.timestamp}>{getTimestamp(entry.timestamp)}</span>
              </div>
              
              <div className={styles.entryContent}>
                <span className={styles.aiIcon}>ü§ñ</span>
                <span className={styles.aiAction}>{entry.chosenAction}</span>
              </div>
              
              {entry.reasoning && (
                <div className={styles.aiReasoning}>
                  <strong>Reasoning:</strong> {entry.reasoning}
                </div>
              )}
              
              {entry.perceivedState && (
                <div className={styles.aiPerceivedState}>
                  <strong>Perceived State:</strong> 
                  <div className={styles.perceivedStateDetails}>
                    <div>Self Health: {entry.perceivedState.self.health}%</div>
                    <div>Enemy Health: {entry.perceivedState.enemy.health}%</div>
                    <div>Self Chi: {entry.perceivedState.self.resources.chi}</div>
                  </div>
                </div>
              )}
              
              {entry.consideredActions && entry.consideredActions.length > 0 && (
                <div className={styles.aiConsideredActions}>
                  <strong>Considered Actions:</strong>
                  <ul>
                    {entry.consideredActions.map((action, actionIndex) => (
                      <li key={actionIndex}>
                        {action.move} (Score: {action.score}) - {action.reason}
                      </li>
                    ))}
                  </ul>
                </div>
              )}
            </div>
          ))
        )}
      </div>
    );
  };

  /**
   * @description Formats all logs for clipboard copying.
   */
  const formatAllLogsForClipboard = () => {
    const battleLogText = battleLog.map(entry => {
      const turn = `T${entry.turn}`;
      const actor = entry.actor;
      const action = entry.action || entry.narrative || entry.result || '';
      const chi = entry.meta && entry.meta.resourceCost !== undefined ? ` (${entry.meta.resourceCost} chi)` : '';
      return `${turn} ${actor}: ${action}${chi}`.trim();
    }).join('\n');
    const aiLogText = aiLog.map(entry => {
      const turn = `T${entry.turn}`;
      const agent = entry.agent;
      const chosen = entry.chosenAction ? `Chose ${entry.chosenAction}` : '';
      const reason = entry.reasoning ? `Reason: ${entry.reasoning}` : '';
      return `${turn} ${agent}: ${chosen}${reason ? '. ' + reason : ''}`.trim();
    }).join('\n');
    return `---\nBATTLE LOG\n---\n${battleLogText}\n\n---\nAI DECISIONS\n---\n${aiLogText}`;
  };

  /**
   * @description Handles copying all logs to clipboard.
   */
  const handleCopyAllLogs = async () => {
    const text = formatAllLogsForClipboard();
    try {
      await navigator.clipboard.writeText(text);
      setCopied(true);
      setTimeout(() => setCopied(false), 1500);
    } catch (err) {
      // fallback: do nothing
    }
  };

  // Disruption state ticker (shows after each turn)
  const renderDisruptionTicker = () => {
    if (!participants.length) return null;
    return (
      <div className={styles.disruptionTicker} aria-label="Disruption State Ticker">
        {participants.map((char) => (
          <span key={char.name} className={`${styles.tickerState} ${styles[char.controlState.toLowerCase()]}`}
            aria-label={`${char.name} Control State: ${char.controlState}, Stability: ${char.stability}, Momentum: ${char.momentum}`}
          >
            <strong>{char.name}</strong>: {char.controlState} | S:{char.stability} | M:{char.momentum}
          </span>
        ))}
      </div>
    );
  };

  return (
    <div className={styles.unifiedBattleLog}>
      {renderDisruptionTicker()}
      <div className={styles.header}>
        <h3>Battle Log</h3>
        <div className={styles.tabSelector}>
          <button
            className={`${styles.tab} ${activeTab === 'narrative' ? styles.active : ''}`}
            onClick={() => setActiveTab('narrative')}
          >
            üìú Narrative ({battleLog.length})
          </button>
          <button
            className={`${styles.tab} ${activeTab === 'ai' ? styles.active : ''}`}
            onClick={() => setActiveTab('ai')}
          >
            ü§ñ AI Decisions ({aiLog.length})
          </button>
        </div>
        <div style={{ display: 'flex', alignItems: 'center', gap: '8px', marginLeft: '16px' }}>
          <input
            type="number"
            placeholder="Turn #"
            min="1"
            max={Math.max(...battleLog.map(e => e.turn), 1)}
            value={turnFilter || ''}
            onChange={(e) => setTurnFilter(e.target.value ? Number(e.target.value) : null)}
            style={{ 
              width: '60px', 
              padding: '4px 8px', 
              borderRadius: 4, 
              border: '1px solid #888', 
              background: '#222', 
              color: '#fff',
              fontSize: '12px'
            }}
          />
          {turnFilter && (
            <button
              onClick={() => setTurnFilter(null)}
              style={{ 
                padding: '2px 6px', 
                borderRadius: 4, 
                border: '1px solid #888', 
                background: '#e74c3c', 
                color: '#fff',
                fontSize: '10px',
                cursor: 'pointer'
              }}
            >
              ‚úï
            </button>
          )}
        </div>
        <button
          onClick={() => setShowAllEntries(!showAllEntries)}
          style={{ marginLeft: 'auto', padding: '6px 14px', borderRadius: 6, border: '1px solid #888', background: showAllEntries ? '#e74c3c' : '#222', color: '#fff', fontWeight: 600, cursor: 'pointer', transition: 'background 0.2s' }}
          title={showAllEntries ? "Show recent entries only (T1 logs will be hidden)" : "Show all entries (including T1 logs)"}
        >
          {showAllEntries ? 'üìú Recent Only' : 'üìú Show All (T1+)'}
        </button>
        <button
          onClick={handleCopyAllLogs}
          style={{ marginLeft: '8px', padding: '6px 14px', borderRadius: 6, border: '1px solid #888', background: copied ? '#27ae60' : '#222', color: '#fff', fontWeight: 600, cursor: 'pointer', transition: 'background 0.2s' }}
          title="Copy all logs to clipboard"
        >
          {copied ? 'Copied!' : 'üìã Copy All Logs'}
        </button>
      </div>
      
      {activeTab === 'narrative' ? renderNarrativeTab() : renderAITab()}
    </div>
  );
}; 