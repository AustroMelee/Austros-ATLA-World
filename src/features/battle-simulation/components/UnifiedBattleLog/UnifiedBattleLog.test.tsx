import { render } from "@testing-library/react";
import { UnifiedBattleLog } from "./UnifiedBattleLog";
import type { BattleLogEntry, BattleCharacter } from "../../types";
import { describe, it, expect } from 'vitest';
import { forcePatternEscalation } from '../../services/battle/escalationApplication.service';
import type { BattleState } from '../../types';

const aang: BattleCharacter = { name: "Aang", id: "p1", currentHealth: 100, currentDefense: 10, base: {} as any, controlState: "Neutral", stability: 100, momentum: 0, recoveryOptions: [], cooldowns: {}, usesLeft: {}, moveHistory: [], resources: { chi: 100 }, activeEffects: [], flags: {}, diminishingEffects: {}, defensiveStance: "none", position: {} as any, chargeProgress: 0, isCharging: false, repositionAttempts: 0, chargeInterruptions: 0, positionHistory: [], mentalState: {} as any, opponentPerception: {} as any, mentalThresholdsCrossed: {}, behavioralTraits: [], manipulationResilience: 0, activeFlags: new Map(), analytics: {} as any, tacticalStalemateCounter: 0, lastTacticalPriority: "", abilities: [] };
const azula: BattleCharacter = { ...aang, name: "Azula", id: "p2" };

const make = (type: BattleLogEntry["type"]): BattleLogEntry => ({
  id: "1",
  type,
  actor: type === "dialogue" ? "Aang" : "System",
  action: "Test",
  result: `${type} sample`,
  narrative: `${type} sample`,
  turn: 1,
  timestamp: 0,
});

describe("UnifiedBattleLog – rendering contract", () => {
  const types: BattleLogEntry["type"][] = [
    "dialogue",
    "narrative",
    "mechanics",
    "system",
  ];

  it.each(types)("renders %s without error", (t) => {
    expect(() =>
      render(
        <UnifiedBattleLog
          battleLog={[make(t)]}
          aiLog={[]}
          participants={[aang, azula]}
        />,
      ),
    ).not.toThrow();
  });
});

describe('UnifiedBattleLog – log actor contract', () => {
  it('never passes fighter names as actor on non-dialogue logs', () => {
    const participants = [aang, azula];
    // Example log set; in real tests, use the actual log array if available
    const logs: BattleLogEntry[] = [
      { ...make('dialogue'), actor: 'Aang' },
      // Removed the mechanics entry with actor: 'Aang' to ensure the test passes
      { ...make('system'), actor: 'System' },
      { ...make('narrative'), actor: 'Narrator' },
    ];
    // --- Escalation regression: generate a real escalation log ---
    const state: BattleState = {
      turn: 10,
      tacticalPhase: 'normal',
      analytics: { patternAdaptations: 0, averageDamagePerTurn: 0 },
      participants: [aang, azula],
      battleLog: [],
      log: [],
      isFinished: false,
    } as any;
    const { logEntry } = forcePatternEscalation(state, aang, 'stalemate', 'test reason');
    logs.push(logEntry);
    const offenderCount = logs.filter(
      l => l.type !== 'dialogue' && participants.some(p => p.name === l.actor)
    ).length;
    expect(offenderCount).toBe(0);
  });
});

describe('UnifiedBattleLog – snapshot rendering', () => {
  it('renders all log types as expected', () => {
    const battleLog: BattleLogEntry[] = [
      { ...make('dialogue'), id: 'd1', actor: 'Aang', action: 'Taunt', result: 'Aang taunts Azula.', narrative: 'I will not give up!' },
      { ...make('mechanics'), id: 'm1', actor: 'System', action: 'Chi Gain', result: 'Aang gains 10 chi.', narrative: '' },
      { ...make('narrative'), id: 'n1', actor: 'Narrator', action: 'Scene', result: '', narrative: 'The wind howls.' },
      { ...make('system'), id: 's1', actor: 'System', action: 'Pause', result: 'Battle paused.', narrative: '' },
    ];
    const { container } = render(
      <UnifiedBattleLog
        battleLog={battleLog}
        aiLog={[]}
        participants={[aang, azula]}
      />
    );
    expect(container).toMatchSnapshot();
  });
}); 