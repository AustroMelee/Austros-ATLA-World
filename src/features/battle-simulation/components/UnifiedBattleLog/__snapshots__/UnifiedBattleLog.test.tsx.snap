// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UnifiedBattleLog â€“ snapshot rendering > renders all log types as expected 1`] = `
<div>
  <div
    class="unifiedBattleLog"
  >
    <div
      aria-label="Disruption State Ticker"
      class="disruptionTicker"
    >
      <span
        aria-label="Aang Control State: Neutral, Stability: 100, Momentum: 0"
        style="margin-right: 12px; font-weight: 500;"
      >
        <strong>
          Aang
        </strong>
        : 
        Neutral
         | S:
        100
         | M:
        0
      </span>
      <span
        aria-label="Azula Control State: Neutral, Stability: 100, Momentum: 0"
        style="margin-right: 12px; font-weight: 500;"
      >
        <strong>
          Azula
        </strong>
        : 
        Neutral
         | S:
        100
         | M:
        0
      </span>
    </div>
    <div
      class="header"
      style="display: flex; align-items: center; gap: 8px; margin-bottom: 8px;"
    >
      <h3
        style="margin: 0px; font-size: 20px;"
      >
        Battle Log
      </h3>
      <div
        class="tabSelector"
        style="display: flex; gap: 4px;"
      >
        <button
          style="font-weight: 700; padding: 4px 10px; border-radius: 4px; border: 1px solid rgb(136, 136, 136); background: rgb(34, 34, 34); color: rgb(255, 255, 255); cursor: pointer;"
        >
          ðŸŽ­ Narrative
        </button>
        <button
          style="font-weight: 400; padding: 4px 10px; border-radius: 4px; border: 1px solid rgb(136, 136, 136); background: rgb(68, 68, 68); color: rgb(255, 255, 255); cursor: pointer;"
        >
          ðŸ’» Technical
        </button>
      </div>
      <div
        style="display: flex; align-items: center; gap: 8px; margin-left: 16px;"
      >
        <input
          max="1"
          min="1"
          placeholder="Turn #"
          style="width: 60px; padding: 4px 8px; border-radius: 4px; border: 1px solid rgb(136, 136, 136); background: rgb(34, 34, 34); color: rgb(255, 255, 255); font-size: 12px;"
          type="number"
          value=""
        />
      </div>
      <button
        style="margin-left: auto; padding: 6px 14px; border-radius: 6px; border: 1px solid rgb(136, 136, 136); background: rgb(231, 76, 60); color: rgb(255, 255, 255); font-weight: 600; cursor: pointer; transition: background 0.2s;"
        title="Show recent entries only (T1 logs will be hidden)"
      >
        ðŸ“œ Recent Only
      </button>
      <button
        aria-label="Copy All Logs"
        style="margin-left: 8px; padding: 6px 14px; border-radius: 6px; border: 1px solid rgb(136, 136, 136); background: rgb(34, 34, 34); color: rgb(255, 255, 255); font-weight: 600; cursor: pointer; transition: background 0.2s;"
        title="Copy all logs to clipboard"
      >
        Copy All Logs
      </button>
      <button
        aria-label="Copy Story Log"
        style="margin-left: 8px; padding: 6px 14px; border-radius: 6px; border: 1px solid rgb(136, 136, 136); background: rgb(34, 34, 34); color: rgb(255, 255, 255); font-weight: 600; cursor: pointer; transition: background 0.2s;"
        title="Copy only the story (narrative) log"
      >
        Copy Story Log
      </button>
      <button
        aria-label="Copy Technical Log"
        style="margin-left: 8px; padding: 6px 14px; border-radius: 6px; border: 1px solid rgb(136, 136, 136); background: rgb(34, 34, 34); color: rgb(255, 255, 255); font-weight: 600; cursor: pointer; transition: background 0.2s;"
        title="Copy only the technical (AI) log"
      >
        Copy Technical Log
      </button>
    </div>
    <div
      style="margin-bottom: 12px;"
    >
      <label
        for="log-view-mode"
        style="color: rgb(203, 213, 225); margin-right: 8px;"
      >
        Log View:
      </label>
      <select
        id="log-view-mode"
      >
        <option
          value="grouped"
        >
          [DEV] Grouped by Turn â€“ Full battle transcript
        </option>
        <option
          value="latest"
        >
          [DEV] Latest Turn Only â€“ Just current turn
        </option>
        <option
          value="full"
        >
          [DEV] All Entries (Raw) â€“ Unfiltered log
        </option>
        <option
          value="public"
        >
          [PUBLIC] Story View â€“ For players/spectators
        </option>
      </select>
    </div>
    <div>
      <div
        style="text-align: left; margin: 0.5em auto 0.5em 0px; padding: 1em; border-radius: 8px; background: rgb(30, 41, 59); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        >
          <div
            style="font-weight: 900; font-size: 20px; margin-bottom: 8px;"
          >
            Aang
          </div>
        </div>
        <div
          style="font-size: 16px;"
        >
          I will not give up!
        </div>
      </div>
      <div
        style="text-align: center; margin: 0.5em auto 0.5em auto; padding: 1em; border-radius: 8px; background: rgb(51, 65, 85); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        />
        <div
          style="font-size: 16px; font-style: italic; white-space: pre-line; word-break: break-word;"
        >
          The wind howls.
        </div>
      </div>
    </div>
    <div
      style="margin-bottom: 24px;"
    >
      <div
        style="font-weight: 700; font-size: 18px; margin: 0px 0px 8px 0px; color: rgb(203, 213, 225); text-align: left;"
      >
        Turn 
        1
      </div>
      <div
        style="text-align: left; margin: 0.5em auto 0.5em 0px; padding: 1em; border-radius: 8px; background: rgb(30, 41, 59); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        >
          <div
            style="font-weight: 900; font-size: 20px; margin-bottom: 8px;"
          >
            Aang
          </div>
        </div>
        <div
          style="font-size: 16px;"
        >
          I will not give up!
        </div>
      </div>
      <div
        style="text-align: center; margin: 0.5em auto 0.5em auto; padding: 1em; border-radius: 8px; background: rgb(51, 65, 85); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        />
        <div
          style="font-size: 16px; font-style: italic; white-space: pre-line; word-break: break-word;"
        />
      </div>
      <div
        style="text-align: center; margin: 0.5em auto 0.5em auto; padding: 1em; border-radius: 8px; background: rgb(51, 65, 85); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        />
        <div
          style="font-size: 16px; font-style: italic; white-space: pre-line; word-break: break-word;"
        >
          The wind howls.
        </div>
      </div>
      <div
        style="text-align: center; margin: 0.5em auto 0.5em auto; padding: 1em; border-radius: 8px; background: rgb(51, 65, 85); color: rgb(255, 255, 255); max-width: 500px; box-shadow: 0 2px 8px rgba(0,0,0,0.08);"
      >
        <div
          style="display: flex; justify-content: space-between; align-items: flex-start;"
        />
        <div
          style="font-size: 16px; font-style: italic; white-space: pre-line; word-break: break-word;"
        />
      </div>
    </div>
  </div>
</div>
`;
